### DISCUSSION TITLE
Bluetooth and SD Card

### DISCUSSION THREAD

**Discussion by 13Pixel** [pschatzmann/arduino-audio-tools#1018] [Q&A] [answered]
Hello, I want to create a project with ESP32 and create something like bluetooth speaker that could play audio from smartphone and SD card. Could I achieve this using this library?

**Comment by pschatzmann** [comment, marked as answer]
Sure, to implement a bluetooth speaker, just look at the A2DP sink examples

Don't forget to read the documentation that can be found in the Readme and Wiki!

**Reply by 13Pixel** [reply]
Thank you

**Comment by GaiVana** [comment]
Hello. I'm trying to do an ESP32 project that takes sound from a Blluetooth or SD card and sends it via I2S to a DSP. When bluetooth is stopped the source is SD card, when SD card is stopped the sound is from bluetooth. Control with buttons. Everything works fine with your library. But I need additional control of indicator and DSP. When I put these additional functions in the main loop, the sound from the SD card cracks. I tried splitting the tasks on different CPU cores. So that the SD card (audio.loop();) works on Core 0, and the additional functions on Core 1. But in this case only SD works. Control functions from another cycle do not work. Is there a possible solution to this problem? I am using Arduino IDE, ESP32-WROOM

**Comment by pschatzmann** [comment]
There are quite a few solution approaches to your problem
- try to keep the delay caused by your additional functions as small as possible!
- try to increase the i2s buffer size/count so that I2S is not running out of data during these delay
- Split up the processing into different tasks and make sure that each task is getting a chance to do some work by calling delay() or if all tasks have the same priority a yield() will do.

By the way you can find this in the [documentation](https://github.com/pschatzmann/arduino-audio-tools/wiki/Introduction#additional-processing-in-the-loop)!

**Reply by GaiVana** [reply]
Thank you. This solved the problem for me. 
To switch the sound from SD card to Bluetooth or vice versa, I use ESP.restart(); and reconfigure similar to the one shown in https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/player-sd_a2dp-audiokit/player-sd_a2dp-audiokit.ino.
Maybe there is another way of switching, without restarting the controller?

**Comment by pschatzmann** [comment]
It depends: if you want to release the full memory and be able to start A2DP again this is the only way.
Just look at the [documentation](https://pschatzmann.github.io/ESP32-A2DP/html/class_bluetooth_a2_d_p_sink.html#a5a91e49987a2e39c09fc6c2a64feaed6) of the end() method in the documentation
